File: tests/output-config.test.ts
import { readFileSync } from 'node:fs';
import { describe, expect, it } from 'vitest';
import { DISPLAY_THEMES } from '../src/display-themes';
import {
  BLIND_BOX_VIEWER_SLOTS_RANGE,
  DISPLAY_FONT_SIZE_RANGE,
  DISPLAY_PANEL_OPACITY_RANGE,
  DISPLAY_SCENE_LAYOUT_IDS,
  GIFT_TARGET_LAYOUT_IDS,
  normalizeBlindBoxDisplayAppearance,
  normalizeDisplayAppearance,
  normalizeDisplaySceneLayout,
  normalizeGiftTargetLayout,
  OUTPUT_CONFIG_SCHEMA_VERSION,
} from '../src/output-config';

const contract = JSON.parse(readFileSync(
  new URL('../testdata/output-config-contract.json', import.meta.url),
  'utf8',
)) as {
  schemaVersion: number;
  displaySceneLayouts: string[];
  giftTargetLayouts: string[];
  displayThemeIds: string[];
  appearance: {
    fontSize: { min: number; max: number; default: number };
    panelOpacity: { min: number; max: number; default: number };
  };
  blindBoxViewerSlots: { min: number; max: number; default: number };
};

const fallback = {
  fontSize: 54,
  accentColor: '#7a4ffb',
  showConnection: true,
  align: 'center' as const,
  panelOpacity: 55,
  defaultDisplayThemeId: 'glass' as const,
};

describe('output configuration contract', () => {
  it('matches the shared TypeScript and Go contract fixture', () => {
    expect(OUTPUT_CONFIG_SCHEMA_VERSION).toBe(contract.schemaVersion);
    expect(DISPLAY_SCENE_LAYOUT_IDS).toEqual(contract.displaySceneLayouts);
    expect(GIFT_TARGET_LAYOUT_IDS).toEqual(contract.giftTargetLayouts);
    expect(DISPLAY_THEMES.map((theme) => theme.id)).toEqual(contract.displayThemeIds);
    expect(DISPLAY_FONT_SIZE_RANGE).toEqual(contract.appearance.fontSize);
    expect(DISPLAY_PANEL_OPACITY_RANGE).toEqual(contract.appearance.panelOpacity);
    expect(BLIND_BOX_VIEWER_SLOTS_RANGE).toEqual(contract.blindBoxViewerSlots);
  });

  it('normalizes layouts and appearances through one interface', () => {
    expect(normalizeDisplaySceneLayout('versus')).toBe('versus');
    expect(normalizeDisplaySceneLayout('unknown')).toBe('stack');
    expect(normalizeGiftTargetLayout('dashboard')).toBe('dashboard');
    expect(normalizeGiftTargetLayout('unknown')).toBe('grid');
    expect(normalizeDisplayAppearance({
      themeId: 'neon', fontSize: 999, accentColor: 'bad', showConnection: false, align: 'right', panelOpacity: 0,
    }, fallback)).toEqual({
      themeId: 'neon', fontSize: 96, accentColor: '#fb7299', showConnection: false, align: 'right', panelOpacity: 10,
    });
    expect(normalizeBlindBoxDisplayAppearance({ viewerSlots: 99 }, fallback)).toEqual(expect.objectContaining({
      viewerSlots: 10,
      themeId: 'glass',
      fontSize: 54,
    }));
  });
});

File: goserver/output_config_test.go
package main

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"reflect"
	"strings"
	"testing"
)

type outputConfigContractFixture struct {
	SchemaVersion       int      `json:"schemaVersion"`
	DisplaySceneLayouts []string `json:"displaySceneLayouts"`
	GiftTargetLayouts   []string `json:"giftTargetLayouts"`
	DisplayThemeIDs     []string `json:"displayThemeIds"`
	Appearance          struct {
		FontSize struct {
			Min     int `json:"min"`
			Max     int `json:"max"`
			Default int `json:"default"`
		} `json:"fontSize"`
		PanelOpacity struct {
			Min     int `json:"min"`
			Max     int `json:"max"`
			Default int `json:"default"`
		} `json:"panelOpacity"`
	} `json:"appearance"`
	BlindBoxViewerSlots struct {
		Min     int `json:"min"`
		Max     int `json:"max"`
		Default int `json:"default"`
	} `json:"blindBoxViewerSlots"`
}

func readOutputConfigContractFixture(t *testing.T) outputConfigContractFixture {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("..", "testdata", "output-config-contract.json"))
	if err != nil {
		t.Fatal(err)
	}
	var fixture outputConfigContractFixture
	if err := json.Unmarshal(data, &fixture); err != nil {
		t.Fatal(err)
	}
	return fixture
}

func TestOutputConfigMatchesSharedContract(t *testing.T) {
	fixture := readOutputConfigContractFixture(t)
	if fixture.SchemaVersion != outputConfigSchemaVersion {
		t.Fatalf("schema version = %d, want %d", outputConfigSchemaVersion, fixture.SchemaVersion)
	}
	for _, layout := range fixture.DisplaySceneLayouts {
		if !isDisplaySceneLayout(layout) {
			t.Fatalf("display scene layout %q is missing", layout)
		}
	}
	if len(displaySceneLayoutIDs) != len(fixture.DisplaySceneLayouts) {
		t.Fatalf("display scene layout count = %d, want %d", len(displaySceneLayoutIDs), len(fixture.DisplaySceneLayouts))
	}
	for _, layout := range fixture.GiftTargetLayouts {
		if !isGiftTargetLayout(layout) {
			t.Fatalf("gift target layout %q is missing", layout)
		}
	}
	if len(giftTargetLayoutIDs) != len(fixture.GiftTargetLayouts) {
		t.Fatalf("gift target layout count = %d, want %d", len(giftTargetLayoutIDs), len(fixture.GiftTargetLayouts))
	}
	for _, themeID := range fixture.DisplayThemeIDs {
		if !isDisplayThemeID(themeID) {
			t.Fatalf("display theme %q is missing", themeID)
		}
	}
	if fixture.Appearance.FontSize.Min != displayFontSizeMin || fixture.Appearance.FontSize.Max != displayFontSizeMax || fixture.Appearance.FontSize.Default != displayFontSizeDefault {
		t.Fatalf("font size contract does not match")
	}
	if fixture.Appearance.PanelOpacity.Min != displayPanelOpacityMin || fixture.Appearance.PanelOpacity.Max != displayPanelOpacityMax || fixture.Appearance.PanelOpacity.Default != displayPanelOpacityDefault {
		t.Fatalf("panel opacity contract does not match")
	}
	if fixture.BlindBoxViewerSlots.Min != blindBoxViewerSlotsMin || fixture.BlindBoxViewerSlots.Max != blindBoxViewerSlotsMax || fixture.BlindBoxViewerSlots.Default != blindBoxViewerSlotsDefault {
		t.Fatalf("viewer slots contract does not match")
	}
}

func TestConfigStorePersistsEveryOutputLayoutAndViewerSlots(t *testing.T) {
	fixture := readOutputConfigContractFixture(t)
	displayScenes := make([]map[string]any, 0, len(fixture.DisplaySceneLayouts))
	for _, layout := range fixture.DisplaySceneLayouts {
		displayScenes = append(displayScenes, map[string]any{
			"id": "scene-" + layout, "name": "Scene " + layout,
			"attributeNames": []string{"Score"}, "layout": layout, "themeId": "glass",
		})
	}
	payload, err := json.Marshal(map[string]any{
		"attributes":    []map[string]any{{"name": "Score", "value": 0, "unit": "none", "format": "number", "decimals": 0, "suffix": ""}},
		"displayScenes": displayScenes,
		"blindBoxDisplay": map[string]any{
			"themeId": "glass", "fontSize": 48, "accentColor": "#fb7299", "showConnection": true,
			"align": "center", "panelOpacity": 55, "viewerSlots": 8,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	store := &configStore{path: filepath.Join(t.TempDir(), "config.json")}
	response := httptest.NewRecorder()
	store.handle(response, httptest.NewRequest(http.MethodPut, "/api/config", strings.NewReader(string(payload))))
	if response.Code != http.StatusOK {
		t.Fatalf("status = %d, want 200; body = %s", response.Code, response.Body.String())
	}
	state, err := store.readState()
	if err != nil {
		t.Fatal(err)
	}
	gotLayouts := make([]string, 0, len(state.DisplayScenes))
	for _, scene := range state.DisplayScenes {
		gotLayouts = append(gotLayouts, scene.Layout)
	}
	if !reflect.DeepEqual(gotLayouts, fixture.DisplaySceneLayouts) {
		t.Fatalf("display scene layouts = %#v, want %#v", gotLayouts, fixture.DisplaySceneLayouts)
	}
	if state.BlindBoxDisplay.ViewerSlots != 8 {
		t.Fatalf("viewer slots = %d, want 8", state.BlindBoxDisplay.ViewerSlots)
	}
	encoded, err := json.Marshal(state.BlindBoxDisplay)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(encoded), `"viewerSlots":8`) || !strings.Contains(string(encoded), `"themeId":"glass"`) {
		t.Fatalf("blind box display JSON = %s", encoded)
	}
}
